src/include/daos_fs_sys.h

```diff
@@ -0,0 +1,506 @@
+	     daos_size_t *size, daos_event_t *ev);
+
+/**
+ * Write data to the file object.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	obj	Opened file object.
+ * \param[in]	buf	Data to write.
+ * \param[in]	off	Offset into the file to write to.
+ * \param[in,out]
+ *		size	[in]: Size of buffer passed in. [out]: Actual size of
+ *			      data written. On error, this is set to -1.
+ * \param[in]	ev	Completion event, it is optional and can be NULL.
+ *			Function will run in blocking mode if \a ev is NULL.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_write(dfs_sys_t *dfs_sys, dfs_obj_t *obj, const void *buf,
+	      daos_off_t off, daos_size_t *size, daos_event_t *ev);
+
+/**
+ * Punch a hole in the file starting at offset to len. If len is set to
+ * DFS_MAX_FSIZE, this will be a truncate operation to punch all bytes in the
+ * file above offset. If the file size is smaller than offset, the file is
+ * extended to offset and len is ignored.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	path	Link path of file.
+ * \param[in]	offset	offset of file to punch at.
+ * \param[in]	len	number of bytes to punch.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_punch(dfs_sys_t *dfs_sys, const char *path,
+	      daos_off_t offset, daos_off_t len);
+
+/**
+ * Similar to dfs_sys_punch but on an open object.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	obj	Opened file object.
+ * \param[in]	offset	offset of file to punch at.
+ * \param[in]	len	number of bytes to punch.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_opunch(dfs_sys_t *dfs_sys, dfs_obj_t *obj,
+	       daos_off_t offset, daos_off_t len);
+
+/**
+ * Remove an object identified by path. If object is a directory and is
+ * non-empty; this will fail unless force option is true. If object is a
+ * symlink, the symlink is removed.
+ * See dfs_remove() for details.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	path	Link path of object.
+ * \param[in]	force	If true, remove dir even if non-empty.
+ * \param[in]	oid	Optionally return the DAOS Object ID of the removed obj.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_remove(dfs_sys_t *dfs_sys, const char *path, bool force,
+	       daos_obj_id_t *oid);
+
+/**
+ * Similar to dfs_sys_remove but optionally enforces a type check
+ * on the entry.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	path	Link path of object.
+ * \param[in]	force	If true, remove dir even if non-empty.
+ * \param[in]	mode	mode_t (S_IFREG | S_IFDIR | S_IFLNK).
+ *			Pass 0 skip the type check.
+ * \param[out]	oid	Optionally return the DAOS Object ID of the removed obj.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_remove_type(dfs_sys_t *dfs_sys, const char *path, bool force,
```

Comment:
what is the use case for this?

Reply:
Really just a rollup of POSIX rmdir, unlink, remove:
- unlink() - removes files, fails otherwise
- rmdir() - removes directories, fails otherwise
- remove() - calls either unlink() or rmdir()